### PR TITLE
Expose ChannelDispatcher and CallbackDebugBehavior.

### PR DESCRIPTION
**Pull request by imcarolwang**
For #4782 .

### CODE REVIEW

File: src/System.Private.ServiceModel/src/System/ServiceModel/Description/CallbackDebugBehavior.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+using System.ServiceModel.Channels;
+using System.ServiceModel.Dispatcher;
+
+namespace System.ServiceModel.Description
+{
+    public class CallbackDebugBehavior : IEndpointBehavior
+    {
+        bool includeExceptionDetailInFaults = false;
+
```
Comment (mconnew) [inline src/System.Private.ServiceModel/src/System/ServiceModel/Description/CallbackDebugBehavior.cs:14]: This is missing accessibility modifier (private), and isn't using the correct naming convention. This field can go away though if you convert IncludeExceptionDetailsInFaults into an auto property.

File: src/System.Private.ServiceModel/src/System/ServiceModel/Description/CallbackDebugBehavior.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+// See the LICENSE file in the project root for more information.
+
+
+using System.ServiceModel.Channels;
+using System.ServiceModel.Dispatcher;
+
+namespace System.ServiceModel.Description
+{
+    public class CallbackDebugBehavior : IEndpointBehavior
+    {
```
Comment (mconnew) [inline src/System.Private.ServiceModel/src/System/ServiceModel/Description/CallbackDebugBehavior.cs:12]: We should have a scenario test for this. You would test this by creating a service which uses a callback contract, throw an exception in a callback method and have the service check that the service side exception that was caught contained the client side exception details.

**Comment by HongGit** [comment]
@imcarolwang can you please review Matt's comments?

**Comment by imcarolwang** [comment]
> @imcarolwang can you please review Matt's comments?

@HongGit , yes I am working on adding test scenario as Matt suggests.

File: src/System.Private.ServiceModel/src/System/ServiceModel/Description/CallbackDebugBehavior.cs (hunk 2 above)
Comment (imcarolwang) [reply src/System.Private.ServiceModel/src/System/ServiceModel/Description/CallbackDebugBehavior.cs:12]: Hi @mconnew , when validating the feature in test scenario, after applying **CallBackDebugBehavior.IncludeExceptionDetailInFaults=true**, the **exception.Message** property value of exception caught from service is `"Exception has been thrown by the target of an invocation."` although the exception.ToString() contains the exception detail. This is different than behavior in Desktop where exception.Message has the exact string initiated by exception thrown from the callback method. Do you have any idea here?

File: src/System.Private.ServiceModel/tests/Scenarios/Client/ExpectedExceptions/CallbackDebugBehaviorTests.cs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        
+        // *** EXECUTE *** \\
+        try
+        {
+            string response = serviceProxy.Hello(greeting, includeExceptionDetailInFaults);
+            Assert.Null(response);
+        }
+        catch
+        {
+            string result = Environment.GetEnvironmentVariable(envVar, EnvironmentVariableTarget.Machine);
```
Comment (imcarolwang) [inline src/System.Private.ServiceModel/tests/Scenarios/Client/ExpectedExceptions/CallbackDebugBehaviorTests.cs:60]: I just realize the environment variable set at service side is not available to client machine. Working on way to validate the test result...

File: src/System.Private.ServiceModel/src/System/ServiceModel/Description/CallbackDebugBehavior.cs (hunk 2 above)
Comment (mconnew) [reply src/System.Private.ServiceModel/src/System/ServiceModel/Description/CallbackDebugBehavior.cs:12]: Can you provide the exact text of what you see on .NET Framework and in Core as I'm not quite understanding what the difference is that you are seeing. This is likely a bug where we aren't unwrapping a TargetInvocationException when getting the exception details.

Comment (imcarolwang) [reply src/System.Private.ServiceModel/src/System/ServiceModel/Description/CallbackDebugBehavior.cs:12]: Ok. If we` throw new Exception("SpeficiedErrorMessage")` in the client call back method, and in service side, the exception is caught via `catch (Exception ex)`, then check the` ex.Message` property , with Desktop client, the value is exact the string "**SpeficiedErrorMessage**", for .NETCore client, it is "**Exception has been thrown by the target of an invocation.**" , the `ex.ToString()` is like below for Core client:

```
System.ServiceModel.FaultException`1[System.ServiceModel.ExceptionDetail]: Exception has been thrown by the target of an invocation. (Fault Detail is equal to An ExceptionDetail, likely created by IncludeExceptionDetailInFaults=true, whose value is:
System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ----> System.Exception: SpeficiedErrorMessage
```

Comment (mconnew) [reply src/System.Private.ServiceModel/src/System/ServiceModel/Description/CallbackDebugBehavior.cs:12]: Yep, I was right. This is a failure to unwrap the TargetInvocationException. I'll see if I can work out where the 1 line of code needs to go to fix it and let you know.

Comment (mconnew) [reply src/System.Private.ServiceModel/src/System/ServiceModel/Description/CallbackDebugBehavior.cs:12]: The issue is caused by calling MethodInfo.Invoke on Core whereas on Framework we use RefEmit to output IL. MethodInfo.Invoke wraps the exception in a TargetInvocationException whereas the RefEmit solution doesn't. The problematic code is at:
https://github.com/dotnet/wcf/blob/72598df55f58b575415ae458f1fd4af332ee0310/src/System.Private.ServiceModel/src/System/ServiceModel/Dispatcher/InvokerUtil.cs#L61-L68

Replace it with the following:
```c#
try
{
    if (returnsValue)
    {
        result = method.[Invoke](https://source.dot.net/System.Private.CoreLib/A.html#c8277ed81262a367)(target, inputsLocal);
    }
    else
    {
        method.[Invoke](https://source.dot.net/System.Private.CoreLib/A.html#c8277ed81262a367)(target, inputsLocal);
    }
}
catch(TargetInvocationException tie)
{
    var actualException = tie.InnerException;
    ExceptionDispatchInfo.Capture(actualException).Throw(); // Keep original call stack
}

Comment (imcarolwang) [reply src/System.Private.ServiceModel/src/System/ServiceModel/Description/CallbackDebugBehavior.cs:12]: Thank you @mconnew ! 
I have replaced the code and test shows the behavior on Core is the same as on Framework now.

**Pull request merged**